rust/chains/hyperlane-cosmos/src/providers/mod.rs

```diff
@@ -116,6 +116,55 @@ impl CosmosProvider {
 
         Ok((account_id, signer_info.sequence))
     }
+
+    /// Calculates the sender and the nonce for the transaction.
+    /// We use `payer` of the fees as the sender of the transaction, and we search for `payer`
+    /// signature information to find the nonce.
+    /// If `payer` is not specified, we use the account which signed the transaction first, as
+    /// the sender.
+    fn sender_and_nonce(&self, tx: &Tx) -> Result<(H256, SequenceNumber), ChainCommunicationError> {
+        let (sender, nonce) = tx
+            .auth_info
+            .fee
+            .payer
+            .as_ref()
+            .map(|payer| self.search_payer_in_signer_infos(&tx.auth_info.signer_infos, payer))
+            .map_or_else(
+                || {
+                    let signer_info = tx.auth_info.signer_infos.get(0).ok_or_else(|| {
+                        ChainCommunicationError::from_other_str("no signer info in default signer")
```

Comment:
We started using chain-specific error structs a while ago where possible. If this error is added as an item under `CosmosError`, we'd get the conversion to `ChainCommunicationError` for free, and could reuse above as well (on line 107)

Reply:
Thank you. Fixed.